Rewrite quite a lot. Travis, Docs, CI stuff to follow.

guyboertje opened the pull request:
Lets discuss this.

Commit:
Rewrite quite a lot. Travis, Docs, CI stuff to follow.

Lets discuss this.

webmat commented on lib/logstash/filters/functions.rb:
```diff
@@ -0,0 +1,105 @@
+    def invalid?(op1, op2)
+      nil
+    end
+  end
```
What's the purpose of `NoDivByZeroCheck`?

If the idea is to always have `invalid?` for all operations, the `NoDivByZero` part of it is irrelevant. Perhaps we can have a `NullCheck` instead, when no check is necessary.

guyboertje replied:
This is a left over - will rename.

webmat commented on lib/logstash/filters/math.rb:
```diff
@@ -19,50 +24,129 @@
+            :plugin => "filter",
+            :type => "math",
+            :error => "Numeric literals are specified as in the calculation but the function invalidates with '#{warning}'. You specified: #{calculation.join(',')}"
+          )
```
Curious: since when do we have `I18n` in Logstash? Although our arbitrary `:error` messages will still be in english, with this. Is there support for a plugin to come with its own internationalization strings?

I haven't been localizing anything, I didn't know we had this.

Sorry, all of this is just chatter, nothing to change here.

guyboertje replied:
I18n was initiated in English but never extended to other languages.AFAIR, it is currently used for messages from start up that are logged to the console. Another reason why removing (J)Ruby will be hard (if i18n is to be kept in the future).

webmat commented on lib/logstash/filters/math.rb:
```diff
@@ -19,50 +24,129 @@
+    case value
+    when Numeric
+      value
+    when LogStash::Timestamp, Time
```
Love this. Time calcs will forever be simplified. Make sure to document it explicitly, though. "you can operate on time fields, they'll be converted to floats, then you can re-parse as a date with UNIX_MS" or some such.

guyboertje replied:
Yes, exactly. I have seen or answered sooo many discuss questions about timestamp calcs - have to use the ruby filter, then the next question is "OK, so how do write the ruby?".
+1 on the docs being comprehensive. That is the main reason why I have not written it yet. Moar feedback!!

webmat commented on lib/logstash/filters/math.rb:
```diff
@@ -19,50 +24,129 @@
+    def validated_literal?() validated; end
+    def inspect() value.inspect; end
+    def to_s() value.to_s; end
+  end
```
I would consider creating a small class instead of a Struct, here. The idea is actually to move the validation/conversion of the value all in one place, inside this class.

The current implementation:
* sets the value without checking it, and the caller method `#register` must set `validated` to true or false. Passing a string + `true` will not actually work, as I understand this. Detecting strings automatically will remove this pitfall.
* Before performing the calculations, `#filter` must know to convert the values first, then hold them somewhere else and do some `nil` management.

Having a small class can simplify all this. Each instance knows whether it got passed a literal or a string, and deals accordingly.

The operator modules could accept instances of this class instead of literal values. Then the operator's validation takes care of the operator-specific validations (e.g. div by 0) and can even take care of validating for generic problems, like missing fields.

Having the generic validation on the operator modules will allow for more holistic error messages than if we check on this small class I'm suggesting.

guyboertje replied:
Will look at it again. The code changed quite a bit when I saw the Complex number result in the power calcs and had to validate for that because up to then it only div by zero was a consideration. I also decided change the code to pre-validate literals to avoid having a literal fall through the validation calls needlessly for every event.
So the code here is a result of those last minute changes.
More eyes on it is a good thing.
While it sits in my editor only one pair of eyes apply. :-)

webmat reviewed (dismissed):
I haven't had time to fully review this, but I love how comprehensive the tests are.

* I added some discussion around implementation details. Some of which may require changes.
* The plugin doesn't have documentation in `index.asciidoc` yet, so this must be added.
* Do you feel like we should conclude the discussion in #3 before shipping this?
  * My opinion is that we're probably ok to release this with the current DSL. If we do this, we must give a clear warning that this is experimental and the DSL is expected to change, potentially without any effort at maintaining backwards compatibility.

guyboertje commented:
As this plugin has not been released yet, we are free to merge this now and decide on array vs string syntax before the release.
It might be possible to support both:
1. the four element array as is current, `[[<op>,<left>,<right>,<target>], [...],[...]]`
2. a hash, `[{target=>"", calculation=>"<left> <op> <right>"}, {...}, {...}]`

I am a big fan of using differentiating classes at initialization or register to prevent doing conditional checks on known settings in the filter method again and again for each event.

I also want to add custom validation to this plugin ala [file input line 195](https://github.com/logstash-plugins/logstash-input-file/blob/master/lib/logstash/inputs/file.rb#L195) and [228](https://github.com/logstash-plugins/logstash-input-file/blob/master/lib/logstash/inputs/file.rb#L228) so that we can produce the validation errors during config validation and not in pipeline building. I created an issue in Logstash (See logstash 9733) to try to ensure that the next gen Java API supports this.

Validating the settings is made on the instance by calling the `config_init` class method before register is called. It would be possible to move the whole register method body to a validator and set the `@calculate_copy` instance variable upon success - it is just more hoops to just through and is potentially fragile regarding the Java API.

Commit:
Move OperandStruct to MathCalulationElements::LiteralElement and
MathCalulationElements::FieldElement for read-only Literals and
read/write Fields
Took the outrageous liberty of
adding MathCalulationElements::RegisterElement too.

guyboertje commented:
@webmat 
Refactored the OperandStruct and added Registers as a reference implementation.

Commit:
Rename register from `R[0]` to `MEM[0]`

Commit:
Add docs, travis, ci folder, and the Round function

jsvd reviewed (approved):
tested manually and code LGTM
I added this project to travis, future prs/commits should trigger builds correctly now
